components/omega/src/ocn/OceanState.cpp

```diff
@@ -0,0 +1,373 @@
+       2,                /// number of dimensions
+       LayerThicknessDim /// dim pointers
+   );
+
+   // Group state metadata
+   auto StateMetaGroup = OMEGA::MetaGroup::create("State");
+
+   Err = StateMetaGroup->addField("NormalVelocity");
+   Err = StateMetaGroup->addField("LayerThickness");
+
+   // Setup IOFields for state variables
+   Err = OMEGA::IOField::define("NormalVelocity");
+   Err = OMEGA::IOField::define("LayerThickness");
+   Err = OMEGA::IOField::attachData<OMEGA::Array2DR8>("NormalVelocity",
+                                                      NormalVelocity[0]);
+   Err = OMEGA::IOField::attachData<OMEGA::Array2DR8>("LayerThickness",
+                                                      LayerThickness[0]);
+
+} // end defineIOFields
+
+//------------------------------------------------------------------------------
+// Destroy parallel decompositions
+void OceanState::finalizeParallelIO() {
+
+   int Err = 0; // default return code
+
+   // Destroy the IO decomp for arrays with (NCells) dimensions
+   Err = IO::destroyDecomp(CellDecompR8);
+   if (Err != 0)
+      LOG_CRITICAL("OceanState: error destroying cell IO decomposition");
+
+   // Destroy the IO decomp for arrays with (NEdges) dimensions
+   Err = IO::destroyDecomp(EdgeDecompR8);
+   if (Err != 0)
+      LOG_CRITICAL("OceanState: error destroying edge IO decomposition");
+
+} // end finalizeParallelIO
+
+//------------------------------------------------------------------------------
+// Read Ocean State
+void OceanState::read() {
+
+   I4 Err;
+
+   // Read LayerThickness
+   int LayerThicknessID;
+   Err = IO::readArray(LayerThicknessH[0].data(), NCellsAll, "layerThickness",
+                       StateFileID, CellDecompR8, LayerThicknessID);
+   if (Err != 0)
+      LOG_CRITICAL("OceanState: error reading layerThickness");
+
+   // Read NormalVelocity
+   int NormalVelocityID;
+   Err = IO::readArray(NormalVelocityH[0].data(), NEdgesAll, "normalVelocity",
+                       StateFileID, EdgeDecompR8, NormalVelocityID);
+   if (Err != 0)
+      LOG_CRITICAL("OceanState: error reading normalVelocity");
+
+} // end read
+
+//------------------------------------------------------------------------------
+// Perform copy to device for state variables
+void OceanState::copyToDevice(int TimeLevel) {
+
+   deepCopy(LayerThickness[TimeLevel], LayerThicknessH[TimeLevel]);
+   deepCopy(NormalVelocity[TimeLevel], NormalVelocityH[TimeLevel]);
+
+} // end copyToDevice
+
+//------------------------------------------------------------------------------
+// Perform copy to host for state variables
+void OceanState::copyToHost(int TimeLevel) {
+
+   deepCopy(LayerThicknessH[TimeLevel], LayerThickness[TimeLevel]);
+   deepCopy(NormalVelocityH[TimeLevel], NormalVelocity[TimeLevel]);
+
+} // end copyToHost
+
+//------------------------------------------------------------------------------
+// Perform time level update
+void OceanState::updateTimeLevels() {
```

Comment:
I am wondering if we should also swap the host arrays here to keep them in sync ?

Reply:
I agree, I'll make that change.